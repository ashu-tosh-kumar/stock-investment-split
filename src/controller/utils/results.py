from tabulate import tabulate

from src.models.financial_instruments import FinancialInstruments

MARGIN_DEVIATION = 0.01


def analyze_and_print_results(financial_instruments: FinancialInstruments, new_investment: float, results: list[float]) -> None:
    """Helper function to analyze and print helpful summary and/or warnings for final results

    Args:
        financial_instruments (FinancialInstruments): Model representing the portfolio
        new_investment (float): New investment to be made
        results (list[float]): Results of the optimizer containing suggested allocation
        of `new_investment` for various instruments
    """
    total_initial_investment: float = sum([investment.initial_investment for investment in financial_instruments.instruments])
    total_final_investment: float = total_initial_investment + new_investment

    allocation_data: list = [
        (
            instrument.identifier,  # Identifier
            ((100 * instrument.initial_investment) / total_initial_investment),  # Initial Investment Ratio
            100 * instrument.target_investment_ratio,  # Target Investment Ratio
            investment,  # Suggested Investment
            ((100 * (instrument.initial_investment + investment)) / total_final_investment),  # Final Ratio
        )
        for instrument, investment in zip(financial_instruments.instruments, results)
    ]

    print(f"Final allocation of investment of unit {new_investment} is as follows:")
    print(tabulate(allocation_data, headers=["Instrument", "Initial Investment %", "Target Investment %", "Suggested Investment", "Final Ratio"]))

    total_suggested_allocation: float = sum(results)
    if abs(total_suggested_allocation - new_investment) / (new_investment) > MARGIN_DEVIATION:
        print(
            f"Total suggested allocation is of unit {total_suggested_allocation} which is more than {MARGIN_DEVIATION*100}% different than investment amount {new_investment}"  # noqa: E501
        )

    print("Please manually verify the results as Mathematics can only aid you but can't absolve you!")
