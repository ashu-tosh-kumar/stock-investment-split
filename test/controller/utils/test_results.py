import sys
from io import StringIO
from test.mocks import DUMMY_FINANCIAL_INSTRUMENTS

from src.controller.utils.results import (
    analyze_and_print_results,
    sanitize_too_small_values,
)


class TestAnalyzeAndPrintResults:
    def test_analyze_and_print_results_should_print_summary_below_deviation(self, snapshot):
        dummy_new_investment = 30
        dummy_results = [10.0, 10.0, 10.0]
        captured_output = StringIO()
        sys.stdout = captured_output

        analyze_and_print_results(DUMMY_FINANCIAL_INSTRUMENTS, dummy_new_investment, dummy_results)
        sys.stdout = sys.__stdout__  # Reset redirect

        assert snapshot == captured_output.getvalue()

    def test_analyze_and_print_results_should_print_summary_above_deviation(self, snapshot):
        dummy_new_investment = 30
        dummy_results = [5.0, 5.0, 5.0]
        captured_output = StringIO()
        sys.stdout = captured_output

        analyze_and_print_results(DUMMY_FINANCIAL_INSTRUMENTS, dummy_new_investment, dummy_results)
        sys.stdout = sys.__stdout__  # Reset redirect

        assert snapshot == captured_output.getvalue()


class TestSanitizeTooSmallValues:
    def test_sanitize_too_small_values_inplace(self, snapshot):
        results = [10.0, 9.876, 10.065]
        inplace = True

        sanitize_too_small_values(results, inplace)

        assert snapshot == results

    def test_sanitize_too_small_values_not_inplace(self, snapshot):
        results = [10.0, 9.876, 10.065]
        inplace = True

        actual_value = sanitize_too_small_values(results, inplace)

        assert snapshot == actual_value
