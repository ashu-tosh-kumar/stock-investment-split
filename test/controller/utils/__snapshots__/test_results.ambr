# serializer version: 1
# name: TestAnalyzeAndPrintResults.test_analyze_and_print_results_should_print_summary_above_deviation
  '''
  Final allocation of investment of unit 30 is as follows:
  Instrument      Initial Investment %    Target Investment %    Suggested Investment    Final %
  ------------  ----------------------  ---------------------  ----------------------  ---------
  INFOSYS                           30                     33                       5    26.9231
                                    50                     33                       5    42.3077
                                    20                     34                       5    19.2308
  Total suggested allocation is of unit 15.0 which is more than 1.0% different than investment amount 30
  Please manually verify the results as Mathematics can only aid you but can't absolve you!
  
  '''
# ---
# name: TestAnalyzeAndPrintResults.test_analyze_and_print_results_should_print_summary_below_deviation
  '''
  Final allocation of investment of unit 30 is as follows:
  Instrument      Initial Investment %    Target Investment %    Suggested Investment    Final %
  ------------  ----------------------  ---------------------  ----------------------  ---------
  INFOSYS                           30                     33                      10    30.7692
                                    50                     33                      10    46.1538
                                    20                     34                      10    23.0769
  Please manually verify the results as Mathematics can only aid you but can't absolve you!
  
  '''
# ---
# name: TestSanitizeTooSmallValues.test_sanitize_too_small_values_inplace
  list([
    10,
    9,
    10,
  ])
# ---
# name: TestSanitizeTooSmallValues.test_sanitize_too_small_values_not_inplace
  list([
    10,
    9,
    10,
  ])
# ---
